libcaf_core/test/config_option.cpp

```diff
@@ -0,0 +1,238 @@
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright (C) 2011 - 2016                                                  *
+ * Dominik Charousset <dominik.charousset (at) haw-hamburg.de>                *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#include "caf/config.hpp"
+
+#define CAF_SUITE config_option
+#include "caf/test/unit_test.hpp"
+
+#include "caf/config_option.hpp"
+#include "caf/actor_system_config.hpp"
+
+using namespace std;
+using namespace caf;
+
+
+namespace {
+
+struct fixture {
+};
+
+constexpr const char* category = "category";
+constexpr const char* name = "name";
+constexpr const char* explanation = "explanation";
+constexpr size_t line = 0;
+
+template<class T>
+constexpr T init_value() { return 0; }
+
+template<class T, class U, class V> 
+constexpr T failure_value() {
+  return static_cast<V>(std::numeric_limits<U>::max()) + 2;
+}
+
+template <class T, class U> 
+std::tuple<T, std::string> run_config_option(const T& init_value, 
+                                             const U& test_value) {
+  std::stringstream ostr;
+  T output_value = init_value;
+  auto cv = config_value(test_value);
+  auto co = make_config_option(output_value, category, name, explanation);
+  auto f = co->to_sink(); 
+  f(line, cv, static_cast<std::ostream&>(ostr));
+  return std::make_tuple(output_value, ostr.str());
+}
+
+template <class T, class U> //only works with an integral types and double 
+void check_num_general_usage() {
+  //check positive numbers
+  std::string error_str;
+  U test_value = 5;
+  T result;
+  std::tie(result, error_str) = run_config_option(init_value<T>(),
+                                                        test_value); 
+  CAF_CHECK_EQUAL(result, test_value);
+  CAF_CHECK_EQUAL(error_str.empty(), true);
+
+  //check negative numbers
+  test_value = -5;
+  std::tie(result, error_str) = run_config_option(init_value<T>(),
+                                                     test_value); 
+  if (std::numeric_limits<T>::is_signed) {
+    CAF_CHECK_EQUAL(result, test_value);
+    CAF_CHECK_EQUAL(error_str.empty(), true);
```

Comment:
Quite verbose to check a boolean expression with `CAF_CHECK_EQUAL`. Why not simply `CAF_CHECK(error_str.empty());`?